### DISCUSSION TITLE
Use Husky in project with multiple micro-frontends

### DISCUSSION THREAD

**Discussion by hagealex** [typicode/husky#1318] [Q&A]
Hello everyone!

I am working on a project that has multiple micro-frontends based on AngularJS. They are all in the same repository and separated by specific folders.

So the project structure looks like this:

Root
  -> micro_frontend_1
  -> micro_frontend_2
  -> micro_fronend_3

What I want to achieve is the following:

When I do a commit I want to run  `ng test` in the directories where I made changes. So not in all folders but only where I made changes.

I know I can specify the pre-commit hook to check for specific file types. But is it also possible to something like I mentioned?

**Comment by typicode** [comment]
Hi,

Most monorepo tools have a way to run tasks only in changed workspaces.
For example: https://turbo.build/repo/docs/core-concepts/monorepos/filtering#filter-by-changed-workspaces

Add husky at the root of the project, create a `<project_root>/.husky/pre-commit` file and inside run you monoropo tool to only run the `test` taks in the changed workspaces.

**Reply by haotian-chen-ooh** [reply]
I am having a similar issue. Is there any way to use husky for multiple subdirectories without installing it in the root directory? Please explain how to do that if it's possible.

I tried install husky for all subdirectories and generating ./husky for all of them, but only 1 directory would trigger pre-commit once a commit is made. It looks like only the latest generated ./husky/pre-commit is triggered, and the others are ignored.

I also checked the document here but it only mentioned the case that only 1 subdirectory has package.json
https://typicode.github.io/husky/how-to.html#project-not-in-git-root-directory

**Reply by typicode** [reply]
Hi,

By design, git allows only for one directory for git hooks. You don't have to put it in the root of the project but you need to script (or use some tools) if you want to run multiple things in different directories.

It can be something simple like:

```shell
# somewhere/precommit
cd path_to_subproject1 && npm something

cd path_to_subproject2 && npm something
```

**Reply by haotian-chen-ooh** [reply]
Thanks for the reply!

Just to confirm. Do you mean instead of installing it in the root directory, we can install husky in only one of the subdirectories, and add the script you mentioned to it?

**Reply by typicode** [reply]
That's correct, see https://typicode.github.io/husky/how-to.html#project-not-in-git-root-directory

The directory containing your hooks can be anywhere. However current working directory is where `.git` is, so you'd have probably to run `cd`.

To test your hook, you can start with:

```sh
# <project>/somedir/pre-commit
pwd # to see where you are

# Your code here

exit 1 # to prevent commit while you're testing your hook
```

**Reply by haotian-chen-ooh** [reply]
That makes sense. Thanks for your help 👍
